### PR TITLE
Unikernel sandbox/container lifecycle implement

### PR DESCRIPTION
**Pull request by Crazykev**
Build based on #219 

1. Implement VM management.
2. Implement container/sandbox lifecycle management.
3. fix some multi-runtime workflow issues in manager

Still lack image management and I need to add some e2e test case.

I've test this manually with critool, with a [memory only redis image](http://downloads.osv.io.s3.amazonaws.com/cloudius/osv-redis-memonly/osv-redis-memonly-v0.24.qemu.qcow2) download from OSv official site. And should store in `/var/lib/frakti/unikernel/image/test-image.qcow2` as test image.

```bash
# cat sandbox.json 
{
    "metadata": {
        "name": "",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "dns_config": {
        "servers": [
               "server1.redhat.com",
               "server2.redhat.com"
         ],
         "searches": [
               "8.8.8.8"
         ]
    },
    "annotations": {
        "runtime.frakti.alpha.kubernetes.io/Unikernel": "true"
    },
    "linux": {
        "security_context": {
            "namespace_options": {
                "host_network": false
            }
        }
    }
}
# cat busybox.json 
{
    "metadata": {
        "name": "busybox"
    },
    "image": {
        "image": "busybox"
    },
    "command": [
        "cat",
	"/etc/resolv.conf"
    ],
    "tty": false,
    "stdin": true
}
# crictl runs sandbox.json
cf198d57-8c21-11e7-8461-4ccc6a36dc71
# crictl sandboxes
SANDBOX ID                             NAME                STATE
cf198d57-8c21-11e7-8461-4ccc6a36dc71                       SANDBOX_READY
# crictl inspects cf198d57-8c21-11e7-8461-4ccc6a36dc71
ID: cf198d57-8c21-11e7-8461-4ccc6a36dc71
UID: hdishd83djaidwnduwk28bcsb
Namespace: default
Attempt: 1
Status: SANDBOX_READY
Created: 2017-08-29 02:50:49.44573354 +0800 CST
Annotations:
	runtime.frakti.alpha.kubernetes.io/Unikernel -> true
# crictl create cf198d57-8c21-11e7-8461-4ccc6a36dc71 busybox.json sandbox.json
08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
# crictl inspect 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
ID: 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
Name: busybox
State: CONTAINER_CREATED
Created: 47 seconds ago
# crictl start 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
ID: 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
Name: busybox
State: CONTAINER_RUNNING
Created: About a minute ago
Started: 3 seconds ago
# crictl stop 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
# crictl inspect 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
ID: 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
Name: busybox
State: CONTAINER_EXITED
Created: About a minute ago
Started: 23 seconds ago
Finished: 2 seconds ago
Exit Code: 0
# crictl stops cf198d57-8c21-11e7-8461-4ccc6a36dc71
ID: cf198d57-8c21-11e7-8461-4ccc6a36dc71
UID: hdishd83djaidwnduwk28bcsb
Namespace: default
Attempt: 1
Status: SANDBOX_NOTREADY
Created: 2017-08-29 02:50:49.44573354 +0800 CST
Annotations:
	runtime.frakti.alpha.kubernetes.io/Unikernel -> true
# crictl rm 08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
08535aec76277c34ea130828dff4390beb275a2efcfde308d3ff8364671ab054
# crictl ps
# crictl rms cf198d57-8c21-11e7-8461-4ccc6a36dc71
Removed sandbox cf198d57-8c21-11e7-8461-4ccc6a36dc71
# crictl sandboxes
```
@feiskyer @resouer PTAL

### CODE REVIEW

**Comment by feiskyer** [comment]
@Crazykev could you fix the CI first?

**Comment by Crazykev** [comment]
@feiskyer Fixed and squashed some commits. PTAL

File: pkg/manager/manager.go [hunk 1]
```diff
@@ -44,6 +44,10 @@ const (
 	OSContainerAnnotationKey = "runtime.frakti.alpha.kubernetes.io/OSContainer"
 	// The annotation value specifying this pod will run by OS container runtime.
 	OSContainerAnnotationTrue = "true"
+	// The annotation key specifying this pod will run by unikernel runtime.
+	UnikernelAnnotationKey = "runtime.frakti.alpha.kubernetes.io/Unikernel"
+	// The annotation value specifying this pod will run by unikernel runtime.
+	UnikernelAnnotationTrue = "true"
```
Comment (feiskyer) [inline pkg/manager/manager.go]: Why do we need this? we should decide the unikernel runtime based on image.

File: pkg/manager/manager.go [hunk 2]
```diff
@@ -320,25 +336,22 @@ func (s *FraktiManager) RemoveContainer(ctx context.Context, req *kubeapi.Remove
...
-	}
-	containers = append(containers, vmContainers...)
 
-	if s.privilegedRuntimeService != nil {
-		osContainers, err := s.privilegedRuntimeService.ListContainers(req.GetFilter())
+	var containers []*kubeapi.Container
+	for _, runtimeService := range []runtime.RuntimeService{
+		s.hyperRuntimeService,
+		s.privilegedRuntimeService,
+		s.unikernelRuntimeService,
```
Comment (feiskyer) [inline pkg/manager/manager.go]: Could we add those runtime services to a list and refer to it when required? (same with other places)

Comment (feiskyer) [reply pkg/manager/manager.go]: Only non-nil runtimes should be kept

File: pkg/unikernel/libvirt/connection.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/connection.go]: Should add a build flag for unikernel runtime, it requires libvirt library for building

File: pkg/unikernel/libvirt/connection.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+// ErrDomainNotFound err means can't find specificed domain
+// returned by domain loop up methods
+var ErrDomainNotFound = errors.New("domain not found")
+
+type LibvirtConnect struct {
+	conn *libvirtgo.Connect
+}
+
+func NewLibvirtConnect(uri string) (*LibvirtConnect, error) {
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/connection.go]: Add documents for exported interfaces and types

File: pkg/unikernel/libvirt/connection.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	glog.V(2).Infof("Defining domain:%q", xml)
+	domain, err := lc.conn.DomainDefineXML(xml)
+	if err != nil {
+		return nil, err
+	}
+	return domain, nil
+}
+
+func (lc *LibvirtConnect) ListDomains() ([]libvirtgo.Domain, error) {
+	domains, err := lc.conn.ListAllDomains(0)
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/connection.go]: Should also filter domains not managed by unikernel runtime

File: pkg/unikernel/libvirt/connection.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return domains, nil
+}
+
+func (lc *LibvirtConnect) LookupDomainByName(name string) (*libvirtgo.Domain, error) {
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/connection.go]: s/LookupDomainByName/GetDomainByName

File: pkg/unikernel/libvirt/connection.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		libvirtErr, ok := err.(libvirtgo.Error)
+		if ok && libvirtErr.Code == libvirtgo.ERR_NO_DOMAIN {
+			return nil, ErrDomainNotFound
+		}
+		return nil, err
+	}
+	return domain, nil
+}
+
+func (lc *LibvirtConnect) LookupDomainByUUIDString(uuid string) (*libvirtgo.Domain, error) {
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/connection.go]: s/LookupDomainByUUIDString/GetDomainByUUID

File: pkg/unikernel/libvirt/vmtools.go [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+	"github.com/golang/glog"
+	libvirtgo "github.com/libvirt/libvirt-go"
+	libvirtxml "github.com/libvirt/libvirt-go-xml"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/frakti/pkg/unikernel/metadata"
+)
+
+const (
+	DefaultMaxMem  = 32768 // size in MiB
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/vmtools.go]: same this file, add documentation

File: pkg/unikernel/libvirt/vmtools.go [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+	DefaultMaxCPUs = 8
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/vmtools.go]: why 8?

File: pkg/unikernel/libvirt/vmtools.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+		return fmt.Errorf("unexpected domain(%q) state(%v) when try to StartVM", domainID, state)
+	}
+
+	// Create domain
+	if err = domain.Create(); err != nil {
+		return fmt.Errorf("failed to create domain(%q): %v", domainID, err)
+	}
+
+	// Check domain state
+	err = wait.PollImmediate(200*time.Millisecond, 10*time.Second, domainRunning(domainID, domain))
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/vmtools.go]: Is 10 seconds too short for nested virtualization (e.g. in dev)?

File: pkg/unikernel/libvirt/vmtools.go [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+			Graphics: []libvirtxml.DomainGraphic{
+				{Type: "vnc", Port: -1},
+			},
+			Videos: []libvirtxml.DomainVideo{
+				{Model: libvirtxml.DomainVideoModel{Type: "cirrus"}},
+			},
+			Controllers: []libvirtxml.DomainController{
+				{Type: "scsi", Index: &scsiControllerIndex, Model: "virtio-scsi"},
+			},
+			// Use this hard coded disk for test
```
Comment (feiskyer) [inline pkg/unikernel/libvirt/vmtools.go]: this is not a test file

File: pkg/unikernel/service/container.go [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+	meta := metadata.ContainerMetadata{
+		ID:        cid,
+		Name:      cName,
+		SandboxID: podSandboxID,
+		Config:    config,
+	}
+
+	// TODO(Crazykev): Prepare container image
+	// use a temp image for test
+	imageRef, err := u.prepareTestImage()
```
Comment (feiskyer) [inline pkg/unikernel/service/container.go]: this is not a test file, do not use test funcs

File: pkg/unikernel/service/sandbox.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+	// Create sandbox metadata.
+	meta := metadata.SandboxMetadata{
+		ID:       podID,
+		Name:     podName,
+		Config:   config,
+		VMConfig: vmMeta,
+	}
+
+	// TODO(Crazykev): Create ns and cni config
```
Comment (feiskyer) [inline pkg/unikernel/service/sandbox.go]: CNI?

File: pkg/unikernel/service/sandbox.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+// sandbox, they should be force terminated.
+func (u *UnikernelRuntime) StopPodSandbox(podSandboxID string) error {
+	sandbox, err := u.sandboxStore.Get(podSandboxID)
+	if err != nil {
+		return fmt.Errorf("failed to find sandbox(%q): %v", podSandboxID, err)
+	}
+	// Stop relate VM
+	if err = u.vmTool.StopVM(sandbox.ID, 0); err != nil {
+		return fmt.Errorf("failed to stop sandbox(%q): %v", sandbox.ID, err)
+	}
```
Comment (feiskyer) [inline pkg/unikernel/service/sandbox.go]: clean up network

File: pkg/unikernel/service/sandbox.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+	return nil
+}
+
+// RemovePodSandbox deletes the sandbox. If there are any running containers in the
+// sandbox, they should be force deleted.
+func (u *UnikernelRuntime) RemovePodSandbox(podSandboxID string) error {
+	// Get sandbox and all containers in sandbox
+	sandbox, err := u.sandboxStore.Get(podSandboxID)
+	if err != nil {
+		if err == store.ErrNotExist {
```
Comment (feiskyer) [inline pkg/unikernel/service/sandbox.go]: Is there any possible of leaking containers here?

File: pkg/unikernel/service/unikernelruntime.go [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+func NewUnikernelRuntimeService(cniNetDir, cniPluginDir, fraktiRoot string, defaultCPU, defaultMem int32) (*UnikernelRuntime, error) {
+	glog.Infof("Initialize unikernel runtime\n")
+
+	// Init VMTools
+	conn, err := libvirt.NewLibvirtConnect(defaultLibvirtdEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(Crazykev): Refactor CNI related code to a common lib.
```
Comment (feiskyer) [inline pkg/unikernel/service/unikernelruntime.go]: CNI refactor deserves a separate PR

File: pkg/manager/manager.go (hunk 1 above)
Comment (Crazykev) [reply pkg/manager/manager.go]: @feiskyer In RunPodSandbox, there is no container image yet, so we need to have some flag to indicate which runtime we are using.

File: pkg/manager/manager.go (hunk 2 above)
Comment (Crazykev) [reply pkg/manager/manager.go]: Yep, that's great idea.

File: pkg/unikernel/libvirt/connection.go (hunk 3 above)
Comment (Crazykev) [reply pkg/unikernel/libvirt/connection.go]: Will do.

File: pkg/unikernel/libvirt/connection.go (hunk 5 above)
Comment (Crazykev) [reply pkg/unikernel/libvirt/connection.go]: For now, I do this when using this domain list(ie. ListContainers) because VMTools does not store anything, although we can filter those by parsing certain pattern domain name here.

File: pkg/unikernel/service/sandbox.go (hunk 15 above)
Comment (Crazykev) [reply pkg/unikernel/service/sandbox.go]: We always try to remove sandbox after removing all inside containers successfully, so I think it is fine here.

File: pkg/unikernel/libvirt/vmtools.go (hunk 9 above)
Comment (Crazykev) [reply pkg/unikernel/libvirt/vmtools.go]: No particular reason, the real limit is defined by kernel through CONFIG_NR_CPUS, this value should be configurable. And to reduce image size and memory cost, some simplified kernel(such as kernel shipped by hyperstart or clear containers) set this num as 8.

File: pkg/unikernel/libvirt/vmtools.go (hunk 10 above)
Comment (Crazykev) [reply pkg/unikernel/libvirt/vmtools.go]: I hope there is no 10s start time unikernel...Do you think 30s enough here?

File: pkg/unikernel/service/container.go (hunk 12 above)
Comment (Crazykev) [reply pkg/unikernel/service/container.go]: Yep, we really need image management here.

File: pkg/unikernel/service/sandbox.go (hunk 13 above)
Comment (Crazykev) [reply pkg/unikernel/service/sandbox.go]: I'm still working on CNI network solution.

File: pkg/manager/manager.go (hunk 1 above)
Comment (feiskyer) [reply pkg/manager/manager.go]: hmm, good catch

**Comment by feiskyer** [comment]
@Crazykev LGTM in general. could you make a rebase?

**Comment by resouer** [comment]
@Crazykev It seems doc is not included? Could you please add one later (or within this PR if you want)?

**Comment by Crazykev** [comment]
@resouer Yep, will include doc in the following one.

**Comment by Crazykev** [comment]
@feiskyer Rebased. PTAL
